extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DataSourceRuntimeConfig.java

```diff
@@ -0,0 +1,39 @@
+package io.quarkus.datasource.runtime;
+
+import java.util.Optional;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class DataSourceRuntimeConfig {
+
+    /**
+     * The datasource username
+     */
+    @ConfigItem
+    public Optional<String> username = Optional.empty();
+
+    /**
+     * The datasource password
+     */
+    @ConfigItem
+    public Optional<String> password = Optional.empty();
+
+    /**
+     * The credentials provider name
+     */
+    @ConfigItem
+    public Optional<String> credentialsProvider = Optional.empty();
+
+    /**
+     * The credentials provider type.
+     * <p>
+     * It is the {@code &#64;Named} value of the credentials provider bean. It is used to discriminate if multiple
+     * CredentialsProvider beans are available.
+     * <p>
+     * For Vault it is: vault-credentials-provider. Not necessary if there is only one credentials provider available.
+     */
+    @ConfigItem
+    public Optional<String> credentialsProviderType = Optional.empty();
```

Comment:
I don't think we need to set the value. We don't do it in any of the other configurations

Reply:
We do. Because I also initialize them via constructor. You cannot trigger the initialization of another map given you have a value present in a map. And when I have some build time config in the map, I need a valid runtime config in the other map so I need to initialize it myself. Talked to David about that a while ago but it's not trivial.

We used this pattern for Flyway IIRC.

Reply:
Understood, thanks